### ISSUE TITLE
Proposal: Remove scope => observable map (observ_id_dict)

### ISSUE THREAD

**Issue by travigd** [JuliaGizmos/WebIO.jl#301]
I'd love to remove `observ_id_dict`. It's stated purpose is for interpolation of observables in JSStrings but that has two issues.
* It's not necessary (observables already have globally unique ids that we can use to refer to them).
* It doesn't work with observables that are in more than one scope*.

* Technically, it does, but only because of the way scopes are globally available in JS. So if I put `obs1` into `scopeA` and `scopeB` (in order), then make a JSString that references `obs1`, it will refer to `scopeB` (even if used with `scopeA`). This still works because we'll look up `scopeB` in the frontend and set the observable there (but it's definitely kind of a hack).

Currently, the behavior is that scopes are containers for observables. I'd like to change that so that scopes instead have _references_ to observables.

Currently the serialization of a scope looks something like this.
```js
scope: {
  dom: { ... },
  imports: [ ... ],
  observables: {
    obs1: {
      id: idOfObs1,
      value: valueOfObs1,
    },
    obs2: { ... },
  },
}
```

I'd like it to look a bit more like this.
```js
scope: {
  dom: { ... },
  imports: [ ... ],
  observables: {
    obs1: idOfObs1,
    obs2: idOfObs2,
  },
```

Then the JS side of things make requests for any observables that aren't currently loaded. This does have the downside of being a little bit more chatty but I don't think it's enough to be noticeable.

The alternative to that (to reduce chattiness) would be to determine all of the observables in the `WebIO.render` phase but I'm also not sure whether or not that's possible. It would also require maintaining a mapping of which clients are subscribed to which observables.

**Comment by shashi** [comment]
So I assume the reason for changing the serialization format is to not serialize the same value twice? I don't know if increasing chattiness instead is going to be better in practice. I'd prefer to leave that as is but remove the mapping. In any case the two features should be separate PRs.

**Comment by travigd** [comment]
Well I don't think there's going to be a way to do that separately. You'd still have to do the roundtrip to check if you have the latest values bc there's no good way to tell if something was just rendered and then displayed (where you wouldn't want to get the latest value - subject to some race conditions) or just refreshed and then displayed (where you would want to get the latest values).

We also already do these round trips for the `setup_scope` messages (which block the kernel in the case of Jupyter at least). I also don't think these are going to be even remotely perceptible (especially since you don't have the normal overhead of creating a new HTTP request which is the usual reason you want to avoid multiple round trips - at least before HTTP 2).

**Comment by travigd** [comment]
I also realize I might not have answered your question.

The reason to change the serialization format is to reflect the way we're conceptualizing the relationship between scopes and observables. Instead of scopes "owning" observables, they have references to them.

**Comment by travigd** [comment]
What I'd like to do is to deprecate the use of `onjs(obs, js"...")` in favor of `onjs(scope, obs, js"...")` (where in the latter, `obs` could either be the `Observable` instance or the name assigned to the observable in that scope).

This would be breaking, but the current behavior is subtly broken anyway. The interpolation only works using the scope that was most recently associated with that observable.

## How It's Broken
```julia
using WebIO

s1 = Scope()
s1(dom"span.date"())
s2 = Scope()
s2(dom"span.date"())

day = Observable("Thursday")
s1["obs"] = day
s2["obs"] = day

# This only defines the JS callback for s2.
# To define one for s1, you have to do onjs(s1, "obs", ...)
onjs(day, js"""
    function(day) {
        this.dom.querySelector(".date").innerText = day;
    }
    """)

display(s1)
display(s2)
```

## Proposal
Only allow the `onjs(scope, obs, callback)` syntax. This is very breaking but I think worth it (maybe for 1.0?). If we don't want to break everything right now, we can keep the current behavior and issue deprecation warnings (but I do want to remove that behavior in the not-too-far-from-now future).

**Comment by piever** [comment]
> What I'd like to do is to deprecate the use of `onjs(obs, js"...")` in favor of `onjs(scope, obs, js"...")` (where in the latter, `obs` could either be the `Observable` instance or the name assigned to the observable in that scope).

I completely agree! For `onjs` to make sense we need to know the `Scope`(to give a value to `_webIOScope`) and the `Observable`. Given that it's problematic for an `Observable` to "belong" to a `Scope`, we need to specify both in the `onjs` signature.

**Comment by shashi** [comment]
> Only allow the onjs(scope, obs, callback) syntax. This is very breaking but I think worth it (maybe for 1.0?). If we don't want to break everything right now, we can keep the current behavior and issue deprecation warnings (but I do want to remove that behavior in the not-too-far-from-now future).

Got it! I think this is good. We should deprecate the current behavior and  have a new method for `onjs`...
